### PR TITLE
Added date checking code; added function to list all charts

### PR DESCRIPTION
**Pull request by johannfrias**
Addresses ideas mentioned in [#38](https://github.com/guoguo12/billboard-charts/issues/38)
 * Added checks for invalid dates (raise `ValueError`)
 * Added checks for dates after the date of a chart's last issue (raise `ValueError`)
 * Added function `list_all_charts` to display all charts on billboard.com (excluding year-end charts), grouped as shown on [https://www.billboard.com/charts](https://www.billboard.com/charts); function is called as `print(billboard.list_all_charts())`

### CODE REVIEW

**Comment by jameswenzel** [comment]
Looks good, but two minor nits:
Can you add a test or two for the value errors? There are already date related tests as well as examples that expect errors. 
Also, in keeping with the style guide in the readme, although it’s not pythonic, renaming the new function to`listAllCharts` would keep the naming consistent. I think @guoguo12 did mention he was skeptical
of scraping chart names, so I’ll defer to him on that. 
Otherwise, looks great! Thanks for putting the work in.

**Comment by guoguo12** [comment]
Great work, @johannfrias! Will review.

> I think @guoguo12 did mention he was skeptical of scraping chart names, so I’ll defer to him on that.

Sorry, to be clear: I like the idea of having a function that scrapes them. I just don't want the function to run behind-the-scenes every time the module is loaded or a `BillboardNotFoundException` is raised.

I agree it should be called `listAllCharts` for consistency. 100% my fault for not using snake case when I created this library. (I was new to Python at the time.)

Alternatively, just rename it to `charts`. So you could ask for `billboard.charts()`.

File: tests/test_dates.py [hunk 1]
```diff
@@ -48,3 +48,13 @@ def test_empty_string_date(self):
...
+
+    @raises(ValueError)
+    def test_invalid_date(self):
+        """Checks that passing a correctly formatted but invalid date raises an exception."""
+        billboard.ChartData('hot-100', date='2018-99-99')
+
+    @raises(ValueError)
+    def test_future_date(self):
+        """Checks that passing a date after the date of a chart's last issue raises an exception."""
+        billboard.ChartData('hot-100', date='3000-10-10')
```
Comment (guoguo12) [inline tests/test_dates.py]: Newline here please ([why?](https://robots.thoughtbot.com/no-newline-at-end-of-file)).

File: billboard.py [hunk 2]
```diff
@@ -273,3 +282,18 @@ def getPositionRowValue(rowName):
...
+    chartGroups = soup.findAll('div', {'class' : "chart-panel__charts"})
+    sFull = ""
+    for category, group in zip(chartCategories, chartGroups):
+        s = category.text.strip()
+        s += '\n' + '-' * len(s) + '\n'
+        charts = group.findAll('a', {'class' : "chart-panel__link"})
+        for c in charts:
+            s += c['href'].split('/')[-1] + '\n'
+        sFull += s + '\n'
+    return sFull
```
Comment (guoguo12) [inline billboard.py]: The current output is very human-readable:

```
Top Charts
----------
hot-100
billboard-200
artist-100
...

Greatest of All Time
--------------------
greatest-billboard-200-albums
greatest-hot-100-singles
greatest-hot-100-artists
...
```

But I would rather the output be easily machine-readable, e.g., a list of strings:

```
['hot-100', 'billboard-200', ...]
```

But that loses the category information, so maybe a dict mapping to lists of strings?

```
{'Top Charts': ['hot-100', 'billboard-200', ...], 'Greatest of All Time': ['greatest-billboard-200-albums', ...]}
```

But that seems too complicated. With a flat list of strings, you can easily get a random chart, search for a chart containing `country`, etc.

@jameswenzel: Any thoughts?

**Comment by guoguo12** [comment]
@johannfrias: For future dates, raising a ValueError seems right to me, because it's exceptional behavior. No one should be be requesting future dates on purpose.

File: billboard.py [hunk 3]
```diff
@@ -179,7 +185,10 @@ def fetchEntries(self):
         dateElement = soup.select_one('button.chart-detail-header__date-selector-button')
         if dateElement:
             dateText = dateElement.text.strip()
-            self.date = datetime.datetime.strptime(dateText, '%B %d, %Y').strftime('%Y-%m-%d')
+            curDate = datetime.datetime.strptime(dateText, '%B %d, %Y')
+            if self.date and curDate < datetime.datetime.strptime(str(self.date), '%Y-%m-%d'):
+                raise ValueError('Date argument is after the date of the latest issue')
```
Comment (guoguo12) [inline billboard.py]: Can you add a comment here (in the code) explaining the reasoning behind this? Something like, "For future dates, Billboard.com returns a valid webpage containing no data but displaying the date of the latest issue."

**Comment by jameswenzel** [comment]
@guoguo12 I like the idea of machine readable, a flat list is probably fine. I don’t think a dict would add much information. 
Also, I just realized: checking the current date isn’t a good indicator of whether or not a date is valid for a chart. Billboard dates charts for the upcoming Saturday.
For example: this week’s Hot 100 was released on Tuesday, but is dated July 28, which is tomorrow. I believe it always has been the following Saturday, so we could check for that. Though then it depends on the day of week the person is using the library... and the chart they are checking (radio charts get released on Monday). I think I remember having this conversation in an old PR 😬
It might be easier to assert that there exists a chart body on 200 responses and assume the absence of one means the user has entered an invalid date. Thoughts?

**Comment by johannfrias** [comment]
Changes:
 * Added the following fixes: [ending newline](https://github.com/guoguo12/billboard-charts/pull/40#discussion-diff-205804713R60), [if-block logic](https://github.com/guoguo12/billboard-charts/pull/40#discussion-diff-205813007R120) and [comment on future dates](https://github.com/guoguo12/billboard-charts/pull/40#discussion-diff-205814086R190)
 * Renamed function from `list_all_charts` to `charts`
 * `charts` now returns the possible charts as a list (the count of which is 126 as of now)

And to comment about future dates, it is indeed a bit tricky, to my understanding: for example, today (7-27) it is possible to request a chart of a future date (7-28) since charts are post-dated. But in two days from now (7-29) it won't be possible to request a chart of the current date (7-29) since the next issue is not released yet.

**Comment by jameswenzel** [comment]
That's correct. On Monday 7/30 you'll be able to request some charts for 8/4, but not others, which will become available Tuesday 7/31. 
I like throwing a value error on an empty response page. If we still want to hardcode in a future date check, we can check that the date isn't >5 days in the future (# of days between Monday and Saturday). Or maybe even count the days until Saturday, with the day of week field. 
Technically, someone can always get the latest chart by not passing in a date, but I don't think it should throw an error if someone does pass in the valid date.

**Comment by jameswenzel** [comment]
One _could_ argue that a datestring for a future date is an "inappropriate value", but I prefer the idea of a more specific Billboard exception.

**Comment by guoguo12** [comment]
Merging this. I'll make a separate issue for raising an error on blank response.

**Comment by guoguo12** [comment]
Thanks again @johannfrias, and apologies for the long delay.

**Pull request merged**